### PR TITLE
Adapt to new YAML-based schema definitions

### LINKED ISSUE #29: Use new schema definition files to avoid duplicating types for e.g. meta
### Description
Now that https://github.com/eiffel-community/eiffel/pull/315 has been merged we should use the new schema definition files to stop generating distinct but not unique types for link objects and meta. Those are now defined in subschemas with the same kind of semantic versioning as the events so we can have a "global" MetaV1 type instead of one per major version of each event type.

### Motivation
This change would allow us to generate fewer types (which helps with the current cluttering) but also make them interchangeable, i.e. a function can act on _any_ link object. We've partially solved this with various interface types but that's not always a realistic option.

### Exemplification
Instead of

```go
type ActivityCanceledV1 struct {
	Data  ActCV1Data  `json:"data"`
	Links ActCV1Links `json:"links"`
	Meta  ActCV1Meta  `json:"meta"`
}
type ActCV1Links []ActCV1Link
type ActCV1Link struct { ... }
```
we get this:
```go
type ActivityCanceledV1 struct {
	Data  ActCV1Data `json:"data"`
	Links LinksV1    `json:"links"`
	Meta  MetaV1     `json:"meta"`
}
type LinksV1 []LinkV1
type LinkV1 struct { ... }
```

### Benefits
See motivation.

### Possible Drawbacks
None.

### PR DESCRIPTION
**Pull request by magnusbaeck**
### Applicable Issues
Fixes #29 

### Description of the Change
Recently, the representation of the schemas in github.com/eiffel-community/eiffel changed (see https://github.com/eiffel-community/eiffel/issues/282) so that schemas and documentation pages are generated from a single YAML source. More importantly, common substructures like `meta` and `links` that de facto have been identical between events from the same commit have been extracted to separate YAML files and referenced from multiple event files.

We hereby adapt the SDK to this new reality and mirror the schema structure in the type structure of the SDK, allowing us to scrap the event-specific structs for `meta` and friends and have single structs shared among all event types.

Note that the PR branch contains multiple commits intended to make it easier to review the changes. I've made each commit message as descriptive as possible. As is customary they will be squashed into a single commit upon merge.

### Alternate Designs
None.

### Benefits
Fewer types, types that can be shared between events, less need to introduce interfaces to be able to write code that can work on any event, easier to write code that acts on e.g. any `meta` field.

### Possible Drawbacks
It's a backwards-incompatible change. Users of the package might not be affected, depending on whether they had references to specific struct versions. For example, code like

```go
e := eiffelevents.NewCompositionDefined()
e.Links = []eiffelevents.CDV3Link{
    eiffelevents.CDV3Link{Target: "...", Type: "..."},
    ...
}
```

would break since the link types have been renamed. That's of course not different from how actual backwards incompatible protocol changes are dealt with, i.e. if we ignore this PR and instead step up the event above to 4.0.0 we'd have to change the link type references to CDV4Link, but if we _do_ merge this PR the same bump to 4.0.0 wouldn't affect the corresponding code _unless_ the reason for the 4.0.0 bump is a backwards incompatible change of the link type. IOW, this is a one time pain (that shouldn't affect most dependents) that overall reduces the frequency with which versioned type references need to be updated.

### Sign-off

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I    have the right to submit it under the open source license    indicated in the file; or

(b) The contribution is based upon previous work that, to the best    of my knowledge, is covered under an appropriate open source    license and I have the right under that license to submit that    work with modifications, whether created in whole or in part    by me, under the same open source license (unless I am    permitted to submit under a different license), as indicated    in the file; or

(c) The contribution was provided directly to me by some other    person who certified (a), (b) or (c) and I have not modified    it.

(d) I understand and agree that this project and the contribution    are public and that a record of the contribution (including all    personal information I submit with it, including my sign-off) is    maintained indefinitely and may be redistributed consistent with    this project or the open source license(s) involved.

Signed-off-by: Magnus Bäck \<magnus.back@axis.com\>

### CODE REVIEW

**Review by t-persson — APPROVED** [review]
This PR does not have any changes from Arica. I assume this is on purpose?

**Comment by magnusbaeck** [comment]
Yes. I'll send a follow-up PR for Arica (which also requires a small code change so we won't attempt to create constants med hyphens and slashes).

File: EiffelEventLinkV1.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+// Add adds a new link of the specified type to a target event identified by an ID.
+func (links *EventLinksV1) AddByID(linkType string, target string) {
+	*links = append(*links, EventLinkV1{Target: target, Type: linkType})
+}
+
+// FindAll returns the IDs of all links of the specified type, or an empty
+// slice if no such links are found.
+func (links EventLinksV1) FindAll(linkType string) []string {
```
Comment (JonasAlfredsson) [inline EiffelEventLinkV1.go:39]: You alternate between (links *EventLinks) and (links EventLinks), is this intentional?

File: internal/cmd/eventgen/linkslice.go [hunk 2]
```diff
@@ -18,49 +18,37 @@ package main
 
 import (
 	_ "embed"
-	"text/template"
+	"fmt"
 
+	"github.com/eiffel-community/eiffelevents-sdk-go"
```
Comment (JonasAlfredsson) [inline internal/cmd/eventgen/linkslice.go:23]: Is this no longer an "internal dependency" and should be on its own like in the previous iteration?

File: internal/cmd/eventgen/schemadefs.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+func (edf *definitionFile) EventType() string {
+	return edf.eventType
+}
+
+func (edf *definitionFile) Filename() string {
+	return edf.filename
+}
+
+func (edf *definitionFile) Version() *semver.Version {
```
Comment (JonasAlfredsson) [inline internal/cmd/eventgen/schemadefs.go:58]: In the two functions above you return a copy of the value, but here you return the pointer. 
Do you want to do this, or is it irrelevant?

**Review by JonasAlfredsson — COMMENTED** [review]
Also, in the "Possible Drawbacks" section you leave us on a bit of a cliffhanger:

> Users of the package might not be affected, depending on whether they had references to specific

SPECIFIC WHAT! I [need](https://media.tenor.com/OWIXSiEmbNwAAAAC/surprise-whats-in-the-box.gif) to know :)

File: EiffelEventLinkV1.go (hunk 1 above)
Comment (magnusbaeck) [reply EiffelEventLinkV1.go:39]: Yes. Not sure if it was done for a good reason in the old code, but this is just a carry-over.

File: internal/cmd/eventgen/linkslice.go (hunk 2 above)
Comment (magnusbaeck) [reply internal/cmd/eventgen/linkslice.go]: Nope, that's an oversight. Correcting commit pushed.

File: internal/cmd/eventgen/schemadefs.go (hunk 3 above)
Comment (magnusbaeck) [reply internal/cmd/eventgen/schemadefs.go:58]: It's a conscious decision mainly based on semver.NewVersion and semver.MustParse both returning *semver.Version (I figured following the pattern of the constructor functions would result in as little trouble as possible). But yes, this is at least a theoretical problem. While most semver.Version methods that perform mutations return a new value, UnmarshalJSON notably doesn't. I don't consider this a problem but I can look into it if you want.

**Comment by magnusbaeck** [comment]
> SPECIFIC WHAT! I [need](https://media.tenor.com/OWIXSiEmbNwAAAAC/surprise-whats-in-the-box.gif) to know :)

Ha! I probably thought "this'll be a boring sentence to write, I'll circle back later" but never did. PR description updated.

**Review by JonasAlfredsson — APPROVED** [review]

**Review by t-persson — APPROVED** [review]

**Comment by magnusbaeck** [comment]
Just resolved a trivial conflict in the last commit so I didn't wait for approvals.

**Pull request merged**